crates/astria-sequencer/src/app/vote_extension.rs

```diff
@@ -0,0 +1,649 @@
+            //
+            // however, if >1/3 of validators are not using the oracle, the prices will not update.
+            bail!("oracle client not set")
+        };
+
+        // if we fail to get prices within the timeout duration, we will return an empty vote
+        // extension to ensure liveness.
+        let prices = match oracle_client.prices(QueryPricesRequest {}).await {
+            Ok(prices) => prices.into_inner(),
+            Err(e) => {
+                bail!("failed to get prices from oracle sidecar: {e}",);
+            }
+        };
+
+        tracing::debug!(
+            prices_count = prices.prices.len(),
+            "got prices from oracle sidecar; transforming prices"
+        );
+
+        let oracle_vote_extension = transform_oracle_service_prices(state, prices)
+            .await
+            .context("failed to transform oracle service prices")?;
+
+        Ok(abci::response::ExtendVote {
+            vote_extension: oracle_vote_extension.into_raw().encode_to_vec().into(),
+        })
+    }
+
+    pub(crate) async fn verify_vote_extension<S: StateReadExt>(
+        &self,
+        state: &S,
+        vote: abci::request::VerifyVoteExtension,
+    ) -> abci::response::VerifyVoteExtension {
+        if vote.vote_extension.is_empty() {
+            return abci::response::VerifyVoteExtension::Accept;
+        }
+
+        match verify_vote_extension(state, vote.vote_extension, false).await {
+            Ok(()) => abci::response::VerifyVoteExtension::Accept,
+            Err(e) => {
+                tracing::error!(error = %e, "failed to verify vote extension");
+                abci::response::VerifyVoteExtension::Reject
+            }
+        }
+    }
+}
+
+// see https://github.com/skip-mev/slinky/blob/5b07f91d6c0110e617efda3f298f147a31da0f25/abci/ve/utils.go#L24
+async fn verify_vote_extension<S: StateReadExt>(
+    state: &S,
+    oracle_vote_extension_bytes: bytes::Bytes,
+    is_proposal_phase: bool,
+) -> anyhow::Result<()> {
+    let oracle_vote_extension = RawOracleVoteExtension::decode(oracle_vote_extension_bytes)
+        .context("failed to decode oracle vote extension")?;
+    let max_num_currency_pairs =
+        DefaultCurrencyPairStrategy::get_max_num_currency_pairs(state, is_proposal_phase)
+            .await
+            .context("failed to get max number of currency pairs")?;
+
+    ensure!(
+        oracle_vote_extension.prices.len() as u64 <= max_num_currency_pairs,
+        "number of oracle vote extension prices exceeds max expected number of currency pairs"
+    );
+
+    for prices in oracle_vote_extension.prices.values() {
+        ensure!(
+            prices.len() <= MAXIMUM_PRICE_BYTE_LEN,
+            "encoded price length exceeded {MAXIMUM_PRICE_BYTE_LEN}"
+        );
+    }
+
+    Ok(())
+}
+
+// see https://github.com/skip-mev/slinky/blob/158cde8a4b774ac4eec5c6d1a2c16de6a8c6abb5/abci/ve/vote_extension.go#L290
+async fn transform_oracle_service_prices<S: StateReadExt>(
```

Comment:
The body of this function should be a stream rather than a loop. Create a future for each (id, price) pair and collect into the `IndexMap` using `TryStreamExt::try_collect` from the futures crate.